File: wisp/wispai/tools.go
package wispai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

// ToolExecutor executes a single tool call and returns a result string.
type ToolExecutor func(ctx context.Context, arguments string) (string, error)

// RegisteredTool pairs a Tool definition with its executor.
type RegisteredTool struct {
	Tool     Tool
	Executor ToolExecutor
}

// NewToolHandler builds a Tool slice and a ToolHandler dispatcher from a list
// of RegisteredTools.
func NewToolHandler(tools []RegisteredTool) ([]Tool, ToolHandler) {
	defs := make([]Tool, len(tools))
	executors := make(map[string]ToolExecutor, len(tools))
	for i, t := range tools {
		defs[i] = t.Tool
		executors[t.Tool.Name] = t.Executor
	}
	return defs, func(ctx context.Context, id, name, arguments string) (string, error) {
		fn, ok := executors[name]
		if !ok {
			return "", fmt.Errorf("unknown tool: %s", name)
		}
		return fn(ctx, arguments)
	}
}

// DefaultTools returns the standard coding tools (bash, read, write, edit,
// grep, find, ls) and a ToolHandler that dispatches to them.
func DefaultTools() ([]Tool, ToolHandler) {
	return NewToolHandler([]RegisteredTool{
		bashTool(),
		readTool(),
		writeTool(),
		editTool(),
		grepTool(),
		findTool(),
		lsTool(),
	})
}

// --- shared constants ---

const (
	outMaxLines = 300
	outMaxBytes = 512 * 1024
)

func truncateTail(s string) string {
	if len(s) > outMaxBytes {
		return "[output truncated]\n" + s[len(s)-outMaxBytes:]
	}
	lines := strings.Split(s, "\n")
	if len(lines) > outMaxLines {
		return "[output truncated]\n" + strings.Join(lines[len(lines)-outMaxLines:], "\n")
	}
	return s
}

func truncateHead(s string, lineLimit int) string {
	if len(s) > outMaxBytes {
		return s[:outMaxBytes] + "\n[output truncated at 512KB]"
	}
	lines := strings.Split(s, "\n")
	if len(lines) > lineLimit {
		return strings.Join(lines[:lineLimit], "\n") + "\n[output truncated]"
	}
	return s
}

func rawSchema(s string) json.RawMessage { return json.RawMessage(s) }

// --- bash ---

func bashTool() RegisteredTool {
	return RegisteredTool{
		Tool: Tool{
			Name:        "bash",
			Description: "Execute a bash command in the current working directory. Returns stdout and stderr combined. Output is truncated to the last 300 lines or 512KB. Optionally provide a timeout in seconds.",
			Parameters: rawSchema(`{
  "type": "object",
  "properties": {
    "command": { "type": "string", "description": "Bash command to execute" },
    "timeout": { "type": "number", "description": "Timeout in seconds" }
  },
  "required": ["command"]
}`),
		},
		Executor: func(ctx context.Context, args string) (string, error) {
			var p struct {
				Command string  `json:"command"`
				Timeout float64 `json:"timeout"`
			}
			json.Unmarshal([]byte(args), &p) //nolint

			cmdCtx := ctx
			if p.Timeout > 0 {
				var cancel context.CancelFunc
				cmdCtx, cancel = context.WithTimeout(ctx, time.Duration(p.Timeout*float64(time.Second)))
				defer cancel()
			}

			cmd := exec.CommandContext(cmdCtx, "bash", "-c", p.Command)
			var out bytes.Buffer
			cmd.Stdout = &out
			cmd.Stderr = &out
			cmd.Run() //nolint — we return output regardless of exit code
			return truncateTail(out.String()), nil
		},
	}
}

// --- read ---

func readTool() RegisteredTool {
	return RegisteredTool{
		Tool: Tool{
			Name:        "read",
			Description: "Read the contents of a file. Output is truncated to 300 lines or 512KB. Use offset and limit to page through large files.",
			Parameters: rawSchema(`{
  "type": "object",
  "properties": {
    "path":   { "type": "string", "description": "Path to the file to read (relative or absolute)" },
    "offset": { "type": "number", "description": "Line number to start reading from (1-indexed)" },
    "limit":  { "type": "number", "description": "Maximum number of lines to read" }
  },
  "required": ["path"]
}`),
		},
		Executor: func(ctx context.Context, args string) (string, error) {
			var p struct {
				Path   string `json:"path"`
				Offset int    `json:"offset"`
				Limit  int    `json:"limit"`
			}
			json.Unmarshal([]byte(args), &p) //nolint

			data, err := os.ReadFile(p.Path)
			if err != nil {
				return "", err
			}

			lines := strings.Split(string(data), "\n")
			start := 0
			if p.Offset > 1 {
				start = p.Offset - 1
			}
			if start >= len(lines) {
				return "", nil
			}
			lines = lines[start:]

			limit := outMaxLines
			if p.Limit > 0 {
				limit = p.Limit
			}
			return truncateHead(strings.Join(lines, "\n"), limit), nil
		},
	}
}

// --- write ---

func writeTool() RegisteredTool {
	return RegisteredTool{
		Tool: Tool{
			Name:        "write",
			Description: "Write content to a file, creating it (and any parent directories) if needed. Overwrites existing files completely.",
			Parameters: rawSchema(`{
  "type": "object",
  "properties": {
    "path":    { "type": "string", "description": "Path to the file to write (relative or absolute)" },
    "content": { "type": "string", "description": "Content to write to the file" }
  },
  "required": ["path", "content"]
}`),
		},
		Executor: func(ctx context.Context, args string) (string, error) {
			var p struct {
				Path    string `json:"path"`
				Content string `json:"content"`
			}
			if err := json.Unmarshal([]byte(args), &p); err != nil {
				return "", err
			}
			if err := os.MkdirAll(filepath.Dir(p.Path), 0o755); err != nil {
				return "", err
			}
			if err := os.WriteFile(p.Path, []byte(p.Content), 0o644); err != nil {
				return "", err
			}
			return fmt.Sprintf("Written %d bytes to %s", len(p.Content), p.Path), nil
		},
	}
}

// --- edit ---

func editTool() RegisteredTool {
	return RegisteredTool{
		Tool: Tool{
			Name: "edit",
			Description: "Edit a file using exact text replacement. Every oldText must be unique in the file. " +
				"All edits are matched against the original file simultaneously — do not emit overlapping edits. " +
				"Merge nearby changes into one call with multiple edits[] entries instead of separate calls. " +
				"Keep oldText as small as possible while still being unique.",
			Parameters: rawSchema(`{
  "type": "object",
  "properties": {
    "path": { "type": "string", "description": "Path to the file to edit" },
    "edits": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "oldText": { "type": "string", "description": "Exact text to replace (must be unique in the original file)" },
          "newText": { "type": "string", "description": "Replacement text" }
        },
        "required": ["oldText", "newText"]
      }
    }
  },
  "required": ["path", "edits"]
}`),
		},
		Executor: func(ctx context.Context, args string) (string, error) {
			var p struct {
				Path  string `json:"path"`
				Edits []struct {
					OldText string `json:"oldText"`
					NewText string `json:"newText"`
				} `json:"edits"`
			}
			if err := json.Unmarshal([]byte(args), &p); err != nil {
				return "", err
			}

			data, err := os.ReadFile(p.Path)
			if err != nil {
				return "", err
			}
			original := string(data)

			type editRange struct {
				start, end int
				newText    string
			}
			ranges := make([]editRange, 0, len(p.Edits))
			for i, e := range p.Edits {
				idx := strings.Index(original, e.OldText)
				if idx == -1 {
					return "", fmt.Errorf("edit %d: oldText not found in file", i+1)
				}
				if strings.Count(original, e.OldText) > 1 {
					return "", fmt.Errorf("edit %d: oldText matches multiple locations (must be unique)", i+1)
				}
				ranges = append(ranges, editRange{idx, idx + len(e.OldText), e.NewText})
			}

			sort.Slice(ranges, func(i, j int) bool { return ranges[i].start < ranges[j].start })
			for i := 1; i < len(ranges); i++ {
				if ranges[i].start < ranges[i-1].end {
					return "", fmt.Errorf("edits %d and %d overlap — merge them into one edit", i, i+1)
				}
			}

			// Apply from end to beginning to preserve positions.
			result := original
			for i := len(ranges) - 1; i >= 0; i-- {
				r := ranges[i]
				result = result[:r.start] + r.newText + result[r.end:]
			}

			if err := os.WriteFile(p.Path, []byte(result), 0o644); err != nil {
				return "", err
			}
			return fmt.Sprintf("Applied %d edit(s) to %s", len(p.Edits), p.Path), nil
		},
	}
}

// --- grep ---

func grepTool() RegisteredTool {
	return RegisteredTool{
		Tool: Tool{
			Name:        "grep",
			Description: "Search file contents for a pattern. Returns matching lines as path:line:content. Output truncated to 100 matches or 512KB.",
			Parameters: rawSchema(`{
  "type": "object",
  "properties": {
    "pattern":    { "type": "string", "description": "Search pattern (regex by default)" },
    "path":       { "type": "string", "description": "Directory or file to search (default: current directory)" },
    "glob":       { "type": "string", "description": "Filter files by glob, e.g. '*.go' or '*.ts'" },
    "ignoreCase": { "type": "boolean", "description": "Case-insensitive search" },
    "literal":    { "type": "boolean", "description": "Treat pattern as literal string, not regex" },
    "limit":      { "type": "number", "description": "Maximum matches to return (default: 100)" }
  },
  "required": ["pattern"]
}`),
		},
		Executor: func(ctx context.Context, args string) (string, error) {
			var p struct {
				Pattern    string `json:"pattern"`
				Path       string `json:"path"`
				Glob       string `json:"glob"`
				IgnoreCase bool   `json:"ignoreCase"`
				Literal    bool   `json:"literal"`
				Limit      int    `json:"limit"`
			}
			json.Unmarshal([]byte(args), &p) //nolint
			if p.Limit == 0 {
				p.Limit = 100
			}
			if p.Path == "" {
				p.Path = "."
			}

			pat := p.Pattern
			if p.Literal {
				pat = regexp.QuoteMeta(pat)
			}
			if p.IgnoreCase {
				pat = "(?i)" + pat
			}
			re, err := regexp.Compile(pat)
			if err != nil {
				return "", fmt.Errorf("invalid pattern: %w", err)
			}

			var sb strings.Builder
			matchCount := 0

			filepath.Walk(p.Path, func(path string, info os.FileInfo, err error) error { //nolint
				if err != nil || info.IsDir() || matchCount >= p.Limit || sb.Len() >= outMaxBytes {
					return nil
				}
				if strings.HasPrefix(filepath.Base(path), ".") {
					if info.IsDir() {
						return filepath.SkipDir
					}
					return nil
				}
				if p.Glob != "" {
					if matched, _ := filepath.Match(p.Glob, filepath.Base(path)); !matched {
						return nil
					}
				}
				data, err := os.ReadFile(path)
				if err != nil {
					return nil
				}
				for i, line := range strings.Split(string(data), "\n") {
					if !re.MatchString(line) {
						continue
					}
					if len(line) > 512 {
						line = line[:512] + "..."
					}
					fmt.Fprintf(&sb, "%s:%d:%s\n", path, i+1, line)
					matchCount++
					if matchCount >= p.Limit {
						break
					}
				}
				return nil
			})

			if matchCount == 0 {
				return "No matches found", nil
			}
			return sb.String(), nil
		},
	}
}

// --- find ---

func findTool() RegisteredTool {
	return RegisteredTool{
		Tool: Tool{
			Name:        "find",
			Description: "Search for files by glob pattern. Supports ** for recursive matching (e.g. '**/*.go', 'src/**/*.ts').",
			Parameters: rawSchema(`{
  "type": "object",
  "properties": {
    "pattern": { "type": "string", "description": "Glob pattern, e.g. '*.go', '**/*.ts', 'src/**/*.spec.ts'" },
    "path":    { "type": "string", "description": "Directory to search in (default: current directory)" },
    "limit":   { "type": "number", "description": "Maximum number of results (default: 1000)" }
  },
  "required": ["pattern"]
}`),
		},
		Executor: func(ctx context.Context, args string) (string, error) {
			var p struct {
				Pattern string `json:"pattern"`
				Path    string `json:"path"`
				Limit   int    `json:"limit"`
			}
			json.Unmarshal([]byte(args), &p) //nolint
			if p.Limit == 0 {
				p.Limit = 1000
			}
			if p.Path == "" {
				p.Path = "."
			}

			var results []string
			filepath.Walk(p.Path, func(path string, info os.FileInfo, err error) error { //nolint
				if err != nil || info.IsDir() || len(results) >= p.Limit {
					return nil
				}
				rel, _ := filepath.Rel(p.Path, path)
				if globMatch(p.Pattern, filepath.ToSlash(rel)) {
					results = append(results, rel)
				}
				return nil
			})

			if len(results) == 0 {
				return "No files found", nil
			}
			sort.Strings(results)
			return strings.Join(results, "\n"), nil
		},
	}
}

// globMatch matches a slash-separated relative path against a glob pattern
// that may contain ** for any number of path components.
func globMatch(pattern, relPath string) bool {
	if !strings.Contains(pattern, "**") {
		matched, _ := filepath.Match(pattern, relPath)
		return matched
	}
	i := strings.Index(pattern, "**")
	prefix := pattern[:i]
	rest := strings.TrimPrefix(pattern[i+2:], "/")

	if prefix != "" {
		if !strings.HasPrefix(relPath, prefix) {
			return false
		}
		relPath = relPath[len(prefix):]
	}
	if rest == "" {
		return true
	}
	// rest must match some suffix of relPath.
	for {
		if matched, _ := filepath.Match(rest, relPath); matched {
			return true
		}
		idx := strings.Index(relPath, "/")
		if idx == -1 {
			return false
		}
		relPath = relPath[idx+1:]
	}
}

// --- ls ---

func lsTool() RegisteredTool {
	return RegisteredTool{
		Tool: Tool{
			Name:        "ls",
			Description: "List directory contents, sorted alphabetically. Directories are suffixed with '/'.",
			Parameters: rawSchema(`{
  "type": "object",
  "properties": {
    "path":  { "type": "string", "description": "Directory to list (default: current directory)" },
    "limit": { "type": "number", "description": "Maximum number of entries (default: 500)" }
  }
}`),
		},
		Executor: func(ctx context.Context, args string) (string, error) {
			var p struct {
				Path  string `json:"path"`
				Limit int    `json:"limit"`
			}
			json.Unmarshal([]byte(args), &p) //nolint
			if p.Limit == 0 {
				p.Limit = 500
			}
			if p.Path == "" {
				p.Path = "."
			}

			entries, err := os.ReadDir(p.Path)
			if err != nil {
				return "", err
			}

			names := make([]string, 0, len(entries))
			for i, e := range entries {
				if i >= p.Limit {
					break
				}
				name := e.Name()
				if e.IsDir() {
					name += "/"
				}
				names = append(names, name)
			}
			return strings.Join(names, "\n"), nil
		},
	}
}

File: wisp-swift/Sources/WispAI/Tools.swift
import Foundation

// MARK: - Tool executor

public typealias ToolExecutor = @Sendable (String) async throws -> String

public struct RegisteredTool {
    public let tool: Tool
    public let executor: ToolExecutor

    public init(tool: Tool, executor: @escaping ToolExecutor) {
        self.tool = tool; self.executor = executor
    }
}

public func newToolHandler(_ tools: [RegisteredTool]) -> ([Tool], ToolHandler) {
    let defs = tools.map(\.tool)
    let map: [String: ToolExecutor] = Dictionary(uniqueKeysWithValues: tools.map { ($0.tool.name, $0.executor) })
    let handler: ToolHandler = { _, name, arguments in
        guard let fn = map[name] else { throw ToolError.unknown(name) }
        return try await fn(arguments)
    }
    return (defs, handler)
}

public func defaultTools() -> ([Tool], ToolHandler) {
    newToolHandler([bashTool(), readTool(), writeTool(), editTool(), grepTool(), findTool(), lsTool()])
}

enum ToolError: Error, LocalizedError {
    case unknown(String)
    var errorDescription: String? {
        if case .unknown(let n) = self { return "unknown tool: \(n)" }
        return nil
    }
}

// MARK: - Helpers

private let outMaxLines = 300
private let outMaxBytes = 512 * 1024

private func truncateTail(_ s: String) -> String {
    var s = s
    if s.utf8.count > outMaxBytes {
        let start = s.utf8.index(s.utf8.endIndex, offsetBy: -outMaxBytes)
        s = "[output truncated]\n" + String(s.utf8[start...])!
        return s
    }
    let lines = s.components(separatedBy: "\n")
    if lines.count > outMaxLines {
        return "[output truncated]\n" + lines.suffix(outMaxLines).joined(separator: "\n")
    }
    return s
}

private func truncateHead(_ s: String, lineLimit: Int = outMaxLines) -> String {
    if s.utf8.count > outMaxBytes {
        return String(s.prefix(outMaxBytes)) + "\n[output truncated at 512KB]"
    }
    let lines = s.components(separatedBy: "\n")
    if lines.count > lineLimit {
        return lines.prefix(lineLimit).joined(separator: "\n") + "\n[output truncated]"
    }
    return s
}

private func schema(_ s: String) -> Data { Data(s.utf8) }

// MARK: - bash

private func bashTool() -> RegisteredTool {
    RegisteredTool(
        tool: Tool(
            name: "bash",
            description: "Execute a bash command in the current working directory. Returns stdout and stderr combined. Output truncated to last 300 lines or 512KB. Optionally provide a timeout in seconds.",
            parameters: schema("""
{
  "type": "object",
  "properties": {
    "command": { "type": "string", "description": "Bash command to execute" },
    "timeout": { "type": "number", "description": "Timeout in seconds" }
  },
  "required": ["command"]
}
""")
        ),
        executor: { args in
            struct P: Decodable { var command: String; var timeout: Double? }
            let p = (try? JSONDecoder().decode(P.self, from: Data(args.utf8))) ?? P(command: args)

            let process = Process()
            process.executableURL = URL(fileURLWithPath: "/bin/bash")
            process.arguments = ["-c", p.command]
            let pipe = Pipe()
            process.standardOutput = pipe
            process.standardError = pipe
            try process.run()

            if let timeout = p.timeout {
                DispatchQueue.global().asyncAfter(deadline: .now() + timeout) {
                    if process.isRunning { process.terminate() }
                }
            }
            let output: String = await withCheckedContinuation { cont in
                process.terminationHandler = { _ in
                    let data = pipe.fileHandleForReading.readDataToEndOfFile()
                    cont.resume(returning: String(data: data, encoding: .utf8) ?? "")
                }
            }
            return truncateTail(output)
        }
    )
}

// MARK: - read

private func readTool() -> RegisteredTool {
    RegisteredTool(
        tool: Tool(
            name: "read",
            description: "Read the contents of a file. Output truncated to 300 lines or 512KB. Use offset and limit to page through large files.",
            parameters: schema("""
{
  "type": "object",
  "properties": {
    "path":   { "type": "string", "description": "Path to the file to read" },
    "offset": { "type": "number", "description": "Line number to start reading from (1-indexed)" },
    "limit":  { "type": "number", "description": "Maximum number of lines to read" }
  },
  "required": ["path"]
}
""")
        ),
        executor: { args in
            struct P: Decodable { var path: String; var offset: Int?; var limit: Int? }
            let p = try JSONDecoder().decode(P.self, from: Data(args.utf8))
            let text = try String(contentsOfFile: p.path, encoding: .utf8)
            var lines = text.components(separatedBy: "\n")
            let start = max(0, (p.offset ?? 1) - 1)
            guard start < lines.count else { return "" }
            lines = Array(lines[start...])
            let limit = p.limit ?? outMaxLines
            return truncateHead(lines.joined(separator: "\n"), lineLimit: limit)
        }
    )
}

// MARK: - write

private func writeTool() -> RegisteredTool {
    RegisteredTool(
        tool: Tool(
            name: "write",
            description: "Write content to a file, creating it (and parent directories) if needed. Overwrites existing files.",
            parameters: schema("""
{
  "type": "object",
  "properties": {
    "path":    { "type": "string", "description": "Path to the file to write" },
    "content": { "type": "string", "description": "Content to write" }
  },
  "required": ["path", "content"]
}
""")
        ),
        executor: { args in
            struct P: Decodable { var path: String; var content: String }
            let p = try JSONDecoder().decode(P.self, from: Data(args.utf8))
            let url = URL(fileURLWithPath: p.path)
            try FileManager.default.createDirectory(at: url.deletingLastPathComponent(),
                                                     withIntermediateDirectories: true)
            try p.content.write(to: url, atomically: true, encoding: .utf8)
            return "Written \(p.content.utf8.count) bytes to \(p.path)"
        }
    )
}

// MARK: - edit

private func editTool() -> RegisteredTool {
    RegisteredTool(
        tool: Tool(
            name: "edit",
            description: "Edit a file using exact text replacement. Each oldText must be unique in the file. All edits are matched against the original file simultaneously — do not emit overlapping edits. Merge nearby changes into one call.",
            parameters: schema("""
{
  "type": "object",
  "properties": {
    "path": { "type": "string", "description": "Path to the file to edit" },
    "edits": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "oldText": { "type": "string", "description": "Exact text to replace (must be unique)" },
          "newText": { "type": "string", "description": "Replacement text" }
        },
        "required": ["oldText", "newText"]
      }
    }
  },
  "required": ["path", "edits"]
}
""")
        ),
        executor: { args in
            struct Edit: Decodable { var oldText: String; var newText: String }
            struct P: Decodable { var path: String; var edits: [Edit] }
            let p = try JSONDecoder().decode(P.self, from: Data(args.utf8))
            var original = try String(contentsOfFile: p.path, encoding: .utf8)

            struct Range { var start: String.Index; var end: String.Index; var newText: String }
            var ranges: [Range] = []
            for (i, edit) in p.edits.enumerated() {
                guard let r = original.range(of: edit.oldText) else {
                    throw EditError.notFound(i + 1)
                }
                let count = original.components(separatedBy: edit.oldText).count - 1
                if count > 1 { throw EditError.notUnique(i + 1, count) }
                ranges.append(Range(start: r.lowerBound, end: r.upperBound, newText: edit.newText))
            }

            // Sort by position and check overlaps
            ranges.sort { $0.start < $1.start }
            for i in 1..<ranges.count {
                if ranges[i].start < ranges[i-1].end { throw EditError.overlap(i, i+1) }
            }

            // Apply from end to beginning
            for r in ranges.reversed() {
                original.replaceSubrange(r.start..<r.end, with: r.newText)
            }
            try original.write(toFile: p.path, atomically: true, encoding: .utf8)
            return "Applied \(p.edits.count) edit(s) to \(p.path)"
        }
    )
}

enum EditError: Error, LocalizedError {
    case notFound(Int)
    case notUnique(Int, Int)
    case overlap(Int, Int)
    var errorDescription: String? {
        switch self {
        case .notFound(let i): return "edit \(i): oldText not found in file"
        case .notUnique(let i, let c): return "edit \(i): oldText matches \(c) locations (must be unique)"
        case .overlap(let i, let j): return "edits \(i) and \(j) overlap — merge them"
        }
    }
}

// MARK: - grep

private func grepTool() -> RegisteredTool {
    RegisteredTool(
        tool: Tool(
            name: "grep",
            description: "Search file contents for a pattern. Returns path:line:content. Output truncated to 100 matches or 512KB.",
            parameters: schema("""
{
  "type": "object",
  "properties": {
    "pattern":    { "type": "string", "description": "Search pattern (regex by default)" },
    "path":       { "type": "string", "description": "Directory or file to search (default: current directory)" },
    "glob":       { "type": "string", "description": "Filter files by glob, e.g. '*.swift'" },
    "ignoreCase": { "type": "boolean", "description": "Case-insensitive search" },
    "literal":    { "type": "boolean", "description": "Treat pattern as literal string, not regex" },
    "limit":      { "type": "number", "description": "Maximum matches to return (default: 100)" }
  },
  "required": ["pattern"]
}
""")
        ),
        executor: { args in
            struct P: Decodable {
                var pattern: String; var path: String?; var glob: String?
                var ignoreCase: Bool?; var literal: Bool?; var limit: Int?
            }
            let p = try JSONDecoder().decode(P.self, from: Data(args.utf8))
            let limit = p.limit ?? 100
            let searchPath = p.path ?? FileManager.default.currentDirectoryPath

            var options: NSRegularExpression.Options = []
            if p.ignoreCase == true { options.insert(.caseInsensitive) }
            let rawPattern = p.literal == true ? NSRegularExpression.escapedPattern(for: p.pattern) : p.pattern
            let re = try NSRegularExpression(pattern: rawPattern, options: options)

            var results: [String] = []
            let enumerator = FileManager.default.enumerator(atPath: searchPath)
            while let relPath = enumerator?.nextObject() as? String {
                guard results.count < limit else { break }
                let fullPath = (searchPath as NSString).appendingPathComponent(relPath)
                var isDir: ObjCBool = false
                FileManager.default.fileExists(atPath: fullPath, isDirectory: &isDir)
                guard !isDir.boolValue else { continue }
                guard !relPath.components(separatedBy: "/").contains(where: { $0.hasPrefix(".") }) else { continue }
                if let glob = p.glob {
                    guard fnmatch(glob, (relPath as NSString).lastPathComponent, 0) == 0 else { continue }
                }
                guard let text = try? String(contentsOfFile: fullPath, encoding: .utf8) else { continue }
                for (i, line) in text.components(separatedBy: "\n").enumerated() {
                    guard results.count < limit else { break }
                    let range = NSRange(line.startIndex..., in: line)
                    if re.firstMatch(in: line, range: range) != nil {
                        let truncLine = line.count > 512 ? String(line.prefix(512)) + "..." : line
                        results.append("\(relPath):\(i + 1):\(truncLine)")
                    }
                }
            }
            return results.isEmpty ? "No matches found" : results.joined(separator: "\n")
        }
    )
}

// MARK: - find

private func findTool() -> RegisteredTool {
    RegisteredTool(
        tool: Tool(
            name: "find",
            description: "Search for files by glob pattern. Supports ** for recursive matching (e.g. '**/*.swift', 'src/**/*.ts').",
            parameters: schema("""
{
  "type": "object",
  "properties": {
    "pattern": { "type": "string", "description": "Glob pattern, e.g. '*.swift', '**/*.json'" },
    "path":    { "type": "string", "description": "Directory to search in (default: current directory)" },
    "limit":   { "type": "number", "description": "Maximum number of results (default: 1000)" }
  },
  "required": ["pattern"]
}
""")
        ),
        executor: { args in
            struct P: Decodable { var pattern: String; var path: String?; var limit: Int? }
            let p = try JSONDecoder().decode(P.self, from: Data(args.utf8))
            let limit = p.limit ?? 1000
            let searchPath = p.path ?? FileManager.default.currentDirectoryPath

            var results: [String] = []
            let enumerator = FileManager.default.enumerator(atPath: searchPath)
            while let relPath = enumerator?.nextObject() as? String {
                guard results.count < limit else { break }
                let fullPath = (searchPath as NSString).appendingPathComponent(relPath)
                var isDir: ObjCBool = false
                FileManager.default.fileExists(atPath: fullPath, isDirectory: &isDir)
                guard !isDir.boolValue else { continue }
                if globMatch(pattern: p.pattern, path: relPath) {
                    results.append(relPath)
                }
            }
            if results.isEmpty { return "No files found" }
            return results.sorted().joined(separator: "\n")
        }
    )
}

private func globMatch(pattern: String, path: String) -> Bool {
    if !pattern.contains("**") {
        return fnmatch(pattern, path, 0) == 0 || fnmatch(pattern, (path as NSString).lastPathComponent, 0) == 0
    }
    let parts = pattern.components(separatedBy: "**")
    let prefix = parts[0]
    var rest = parts.dropFirst().joined(separator: "**")
    if rest.hasPrefix("/") { rest = String(rest.dropFirst()) }

    var remaining = path
    if !prefix.isEmpty {
        guard remaining.hasPrefix(prefix) else { return false }
        remaining = String(remaining.dropFirst(prefix.count))
    }
    guard !rest.isEmpty else { return true }

    while true {
        if fnmatch(rest, remaining, 0) == 0 { return true }
        guard let slash = remaining.firstIndex(of: "/") else { return false }
        remaining = String(remaining[remaining.index(after: slash)...])
    }
}

// MARK: - ls

private func lsTool() -> RegisteredTool {
    RegisteredTool(
        tool: Tool(
            name: "ls",
            description: "List directory contents, sorted alphabetically. Directories are suffixed with '/'.",
            parameters: schema("""
{
  "type": "object",
  "properties": {
    "path":  { "type": "string", "description": "Directory to list (default: current directory)" },
    "limit": { "type": "number", "description": "Maximum number of entries (default: 500)" }
  }
}
""")
        ),
        executor: { args in
            struct P: Decodable { var path: String?; var limit: Int? }
            let p = (try? JSONDecoder().decode(P.self, from: Data(args.utf8))) ?? P()
            let limit = p.limit ?? 500
            let dir = p.path ?? FileManager.default.currentDirectoryPath
            let url = URL(fileURLWithPath: dir)
            let entries = try FileManager.default.contentsOfDirectory(
                at: url, includingPropertiesForKeys: [.isDirectoryKey], options: []
            )
            let names: [String] = entries.prefix(limit).map { entry in
                let isDir = (try? entry.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory == true
                return entry.lastPathComponent + (isDir ? "/" : "")
            }.sorted()
            return names.joined(separator: "\n")
        }
    )
}
